optimum/habana/transformers/trainer.py

```diff
@@ -181,7 +190,7 @@ def __init__(
                 # To avoid warnings about parallelism in tokenizers
                 os.environ["TOKENIZERS_PARALLELISM"] = "false"
 
-            if self.gaudi_config.use_habana_mixed_precision:
+            if self.gaudi_config.use_habana_mixed_precision and not self.use_hpu_amp:
```

Comment:
We can remove `and not self.use_hpu_amp` here and raise an error just before if both `self.gaudi_config.use_habana_mixed_precision` and `self.use_hpu_amp` are True (the same check as in gaudi_configuration.py).

Reply:
We can't raise any exceptions as it would be impossible to use autocast. This if statement is responsible for converting ops using hmp. Given that user hasn't specified to use autocast, decision to cast using hmp is left to HabanaConfig.

Reply:
Not sure I understand. Basically what I'm proposing here is the following:
```py
if self.gaudi_config.use_habana_mixed_precision and (args.bf16 or self.gaudi_config.use_torch_autocast):
    raise ValueError(...)

[...]

if self.gaudi_config.use_habana_mixed_precision:
    hmp.convert(...)
``` 
The user can enable autocast with `args.bf16` or `gaudi_config.use_torch_autocast` (the first one being easier to use), but these are not compatible with `gaudi_config.use_habana_mixed_precision`.

Reply:
Yes but in this case user first needs to download and alter/create their own gaudi_config.json instead of using preexisting ones. In  what I propose hpu_precision_backend argument takes precedence over gaudi_config.use_habana_mixed_precision. Args.bf16 is not easier to use as it also requires the change of gaudi_config.json for now. Of course it is not the case if you will change all configs use_habana_mixed_precision to false but I have no information of how are those configuration files going to be altered.

Reply:
I admit that usage of gaudi_config.use_torch_autocast seems kind of strange in this scenario. I thought of it rather as a temporary workaround to control whether autocast is enabled for specific models as changes in GaudiTrainer are model agnostic.

Reply:
I agree that bf16 would resemble Transformers more but then we would use auto backend which looks at the device used. I don't believe any of those two approaches is perfect (either bf16 or half_precision_backend). My line of reasoning looks like this. Imagine there is no GaudiTrainer and you can use models on Gaudi with plain Transformers library. What would you need to do to run mixed precision training? Set bf16 which is the same as it was. You also need to specify half_precision_backend and yes it is inferred from device in case of auto but specifying it seems more verbose. Idk it was just the gut feeling or sth like it. I'm not really attached to my solution, only thing is that I think we should either use bf16 or half_precision_backend not both, your call :)